### ISSUE TITLE
Client functions assume endpoint paths are relative to the page being served

### ISSUE THREAD

**Issue by 3noch** [imalsogreg/servant-reflex#3]
If you serve an API from `localhost/api` and serve a Reflex-DOM page from `localhost/static`, the client functions will try to access the API at `localhost/static/api` instead of `localhost/api`. This precludes people from serving their Reflex-DOM pages from any path that is not a subset of the API paths.

**Comment by imalsogreg** [comment]
I believe this is solved by setting the BasePath explicitly to "/". I haven't written that part of the library yet. I think I can get to it tomorrow though - I don't expect that part to be hard.

FYI - base path and response headers are the last two major missing parts (well - tests would be nice too). The response headers will be a bit harder as reflex-dom's XhrResponse record doesn't have a field for the headers.

**Comment by 3noch** [comment]
Cool! I thought I had tried that, but maybe I missed something. I'll try again.

Do you think many changes will be necessary for servant 0.6? The default client code changed, but this library isn't using a `Manager` anyway.

**Comment by imalsogreg** [comment]
Don't try it again - you'll be disappointed! :) The BaseUrl parameter is settable, but I don't actually use it yet in building the requests - the place to fill in the implementation is [here](https://github.com/mightybyte/servant-reflex/blob/56970bc3efba45531ef8fce8a3a0bfef584e7ade/src/Servant/Common/Req.hs#L102)

I don't know of what's in the works for 0.6, but I was under the impression that servant hq wanted to make smaller, more frequent releases, so perhaps servant-reflex won't need many changes. Is there any particular servant 0.6 feature you're thinking of?

**Comment by 3noch** [comment]
Thanks for looking into it!

servant 0.6 is already in it`s `master` branch. It's just that they got rid of the need to pass `Manager` to the `client` function. I don't know if any architectural things changed that would require some changes in `servant-reflex`.

**Comment by Ikyushii** [comment]
> The BaseUrl parameter is settable, but I don't actually use it yet in building the requests

Oh, that would explain why my tests keep ignore the BaseUrl port I gave to them.

I was wondering if I could help. Unfortunatly, I’m quite new at both servant and reflex, but who knows. I would love to get involve in servant-reflex, as I plan to use it in my project.

**Comment by imalsogreg** [comment]
@Ikyushii I would love your help. Whether you wanted to try integrating BaseUrl into the requests by following the [pattern used in Req.hs](https://github.com/imalsogreg/servant-reflex/blob/30af98a2f232521914990b0d404e20d5f7c5c369/src/Servant/Common/Req.hs#L97-L141) for assembling requests, or to try something simpler (improving the example application, writing some tests, anything else)... any contributions are welcome :) Feel tree to ask for help if you get stuck, here or on freenode #reflex-frp

**Comment by Ikyushii** [comment]
Count me in, I will try the BaseUrl thing this week, probably tomorrow. It might be a good way to understand reflex internals :).

**Comment by Ikyushii** [comment]
Okay so just to let you know, I worked on it today and it almost works. I just have a little edge case to handle and I will be able to do a PR (:

If you want to have a look: https://github.com/Ikyushii/servant-reflex/commit/dd13dc2a7a7478df86c82b4dd5c922e2ff18bcf3

**Comment by 3noch** [comment]
@imalsogreg Glad this is getting worked on. Sorry for not being more help ATM. I got lambasted recently and haven't time.

@Ikyushii Did you consider making host and port portions of `BaseUrl` optional? In that case the XML request would be sent to the same server that served the page. That could be useful as well.

**Comment by Ikyushii** [comment]
The BaseUrl is not defined within servant-reflex but in servant itself, so it is not possible as-is.

**Comment by 3noch** [comment]
I thought the same thing, but: https://github.com/imalsogreg/servant-reflex/blob/master/src/Servant/Common/BaseUrl.hs#L30

**Comment by Ikyushii** [comment]
I think this file can be removed, doesn’t it?

Rather than modifying an existing data type, maybe we can provide helper functions. For instance, `fromCurrentDomain :: String -> BaseUrl`. What do you think?

**Comment by imalsogreg** [comment]
@3noch @lethom I tried defining BaseUrl as this:

```haskell
data BaseUrl t = BaseUrl
  { baseUrlServer :: Behavior t (Maybe (Scheme,String, Maybe Int))
  , baseUrlPath   :: Behavior t String
  }
```

So the server part is all optional (leaving it out will default to a same-host kind of URL, with an optional port inside, and all separate from the path part.

The full [diff is here](https://github.com/imalsogreg/servant-reflex/compare/base-url-2?expand=1)

I put together an example to help debugging: http://web.mit.edu/greghale/Public/example.jsexe/

What do you think about this direction?

**Comment by lethom** [comment]
As I said, I’d rather use the type `Either BaseUrl Sting`. That being said, the fact that BaseUrl is no longer a Dynamic is, I think, a good thing.

**Comment by imalsogreg** [comment]
closed by 43e21802b31b026a7a1fdabe27484ccd7d5d3486
